### PR TITLE
feat(ui): Make named interface elements generate fake boxes

### PR DESCRIPTION
**Pull request by tibetiroka**
**Feature**

This PR partially addresses the feature described in https://github.com/endless-sky/endless-sky/pull/10187#discussion_r1813228643

## Summary
`Interface::GetBox()` now returns the bounding box of a **named** `Interface::Element`, if there is no box with that name. This should make it unnecessary to create boxes and elements individually for every clickable interface element.

Please note that currently only bars are named; sprites use the sprite name as a "name", which is the exact opposite of a stable interface name, so it's not included in this feature.

In the future, the syntax of other interface elements can be expanded with a name, and they will be able to use this feature without extra work.

## Testing Done
Not yet.

## Wiki Update
N/A

### CODE REVIEW

File: source/Interface.cpp [hunk 1]
```diff
@@ -383,6 +389,15 @@ Rectangle Interface::Element::Bounds(const Information &info) const
 
 
 
+// The name of this element, or an empty string if it doesn't have a name.
+std::string Interface::Element::Name() const
```
Comment (TomGoodIdea) [inline source/Interface.cpp:393]:
```suggestion
string Interface::Element::Name() const
```

File: source/Interface.cpp [hunk 2]
```diff
@@ -383,6 +389,15 @@ Rectangle Interface::Element::Bounds(const Information &info) const
...
+{
+	return name;
+}
+
+
+
+
```
Comment (TomGoodIdea) [inline source/Interface.cpp:400]:
```suggestion
```

**Review by TomGoodIdea — APPROVED** [review]
Maybe it would be useful to add a warning if two elements (especially of different types) have the same name? (It can be done in a future PR adding support for more element types.)

**Comment by tibetiroka** [comment]
That would be interesting, though that would require knowing the type of an element, which is non-trivial in this case... we'll figure it out when it comes up, I suppose.

File: source/Interface.cpp [hunk 3]
```diff
@@ -180,7 +182,11 @@ Rectangle Interface::GetBox(const string &name) const
 {
 	auto it = points.find(name);
 	if(it == points.end())
-		return Rectangle();
+	{
+		auto it2 = namedElements.find(name);
+		if(it2 != namedElements.end())
+			return it2->second->Bounds();
+	}
```
[the remark this answers is not in the document]
Comment (TomGoodIdea) [reply source/Interface.cpp:189]: depends if we're doing this every frame

[the remark this answers is not in the document]
Comment (tibetiroka) [reply source/Interface.cpp:189]: I suppose... though it also depends on the ratio of real to fake box lookups, as looking up actual boxes would take longer.

[the remark this answers is not in the document]
Comment (Koranir) [reply source/Interface.cpp:189]: > I don't think a second search is that bad?

rbtree lookup is log n, so putting more elements into one map uses less iterations average than two separate lookups.

**Comment by tibetiroka** [comment]
This is probably too junky of a solution in the long run. I'm not sure how to do better without a significantly larger refactor, though.

**Pull request closed without merging**